lib/offline.js

```diff
@@ -55,60 +61,121 @@ module.exports = function(config) {
       }
     }
 
-    resolve(ghslug('./').catch(function() {
+    ghslug('./')
+    .catch(function() {
       // Use the name from package.json if there's an error while fetching the GitHub slug.
       try {
         return JSON.parse(fs.readFileSync('package.json')).name;
       } catch (ex) {
         return '';
       }
-    }).then(function(cacheId) {
-      var staticFileGlobs = fileGlobs.map(function(v) {
+    })
+    .then(function(cacheId) {
+      checkImports(importScripts, rootDir);
+
+      var absoluteGlobs = fileGlobs.map(function (v) {
         return path.join(rootDir, v);
       });
+      var files = flatGlobs(absoluteGlobs);
+      var filesAndHashes = getFilesAndHashes(files, rootDir, absoluteGlobs);
+
+      var replacements = {
+        cacheId: cacheId,
+        ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
+        cacheVersion: getHashOfHashes(pluckHashes(filesAndHashes)),
+        resources: filesAndHashes,
+        importScripts: importScripts,
+      };
 
-      staticFileGlobs.forEach(function(globPattern) {
-        glob.sync(globPattern.replace(path.sep, '/')).forEach(function(file) {
-          var stat = fs.statSync(file);
+      var stream = gulp.src([__dirname + '/../templates/app/offline-worker.js'])
+        .pipe(template(replacements))
+        .pipe(gulp.dest(rootDir));
 
-          if (stat.isFile() && stat.size > 2 * 1024 * 1024 && staticFileGlobs.indexOf(file) === -1) {
-            console.log(chalk.yellow.bold(file + ' is bigger than 2 MiB. Are you sure you want to cache it? To suppress this warning, explicitly include the file in the fileGlobs list.'));
-          }
-        });
+      return new Promise(function (resolve, reject) {
+        stream.on('finish', function () { resolve(); });
+        stream.on('error', function (e) { reject(e); });
```

Comment:
Nit: no need to define another function, simply do `stream.on('finish', resolve);`